### PR TITLE
RavenDB-21141 Free list handling for containers in the case of rapidl…

### PR DESCRIPTION
**Pull request by ayende**
…y switching between page-level-metadata items (as we now do with stored fields)

### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21141 
https://issues.hibernatingrhinos.com/issue/RavenDB-21140

### Additional description

We had a bunch of issues where we wouldn't use the free list properly for containers. In particular, when switching between different fields during indexing, we would allocate a page very quickly and not reuse them.

This change dropped a 1GB index to < 16MB index!

This fixes the issue, but we already touch this same area here:
https://github.com/ravendb/ravendb/pull/17047

That would _probably_ fix some of the problems, but not all.

There is also a potential to _remember_ free list values per container per page level metadata, but that is something that I would rather do after the https://github.com/ravendb/ravendb/pull/17047 work is done.

Should be merged and cause a conflict (which we'll resolve later), because of the impact.

### CODE REVIEW

**Comment by arekpalinski** [comment]
@redknightlois I'm merging it for testing since tests looks fine. Please still do the review.

File: src/Voron/Data/Containers/Container.cs [hunk 1]
```diff
@@ -441,32 +448,27 @@ private static Container MoveToNextPage(LowLevelTransaction llt, long containerI
             // This is the case where at some point we need to just give up or end up wasting more time to find a page than the time
             // we will use to create and store in disk a new one.
             int i = 0;
+            using var it = ScanFreePages(GetFreeListTree(llt, rootContainer)).GetEnumerator();
             for (; i < tries; i++)
```
Comment (redknightlois) [inline src/Voron/Data/Containers/Container.cs:452]: Is this a high traffic performance critical section?

File: src/Voron/Data/Containers/Container.cs [hunk 2]
```diff
@@ -499,12 +501,39 @@ void RemovePageFromContainerFreeList(Container parent, Container page)
...
+            
+            void AddPageToContainerAllPagesList(Container parent, Container page)
             {
                 page.Header.OnFreeList = true;
                 ModifyMetadataList(llt, parent, ContainerPageHeader.AllPagesOffset, add: true, page.Header.PageNumber);
             }
         }
 
+        private static IEnumerable<long> ScanFreePages(Tree freeList)
+        {
```
Comment (redknightlois) [inline src/Voron/Data/Containers/Container.cs:517]: If we are in a high traffic section, I would rather do this without `IEnumerable<long>` and use a `struct` iterator.

File: src/Voron/Data/Containers/Container.cs (hunk 1 above)
Comment (ayende) [reply src/Voron/Data/Containers/Container.cs:452]: Yes, hopefully fixed in https://github.com/ravendb/ravendb/pull/17047

**Pull request merged**